enhancements/csvless-bundles.md

```diff
@@ -24,209 +24,488 @@ status: implementable
+Add a new command to `opm`, `opm bundle validate`, which takes as input an Operator Bundle and fails with informative errors should the bundle be invalid.
+
+```sh
+$ opm bundle validate . # validate the bundle in the current directory, printing the results to stdout
+$ opm bundle validate image://quay.io/my/bundle:v1.0.0 -o results.xml # pull, unpack, and validate the remote bundle image, printing the results to a file
+```
 
-Supporting indexes with different mediatypes is a key requirement. Both `registry+v2`  and `registry+v1` type bundles should be able to be added to an existing index. There is the potential of versioning the index itself to help differentiate between newer index images, that include the `registry+v2` mediatype, and older versions.
+Results are in [XUnit format](todo).
 
-#### Resolving, Unpacking, and Applying
+### Risks and Mitigations
+
+One risk is around backporting indexes with mixed mediatypes to older clusters. This needs to be supported to ensure upgrades and downgrades of OpenShift behave as intended. This is outlined further in the Upgrade/Downgrade section below.
+
+Another risk is around operatorhub.io - it will need to be adjusted to support the `registry+v2` bundles. This work is not scoped as part of this enhancement and is done by a seperate team - how much effort is involved there is unclear.
+
+Security impacts are minimal because the number of kubernetes objects that OLM supports is still the same - just the delivery mechanism (the bundle format) is changing.
+
+## Design Details
+
+### Implementation Details
+
+
+#### Registry Database Schema Changes
+
+In order to support adding `registry+v2` bundles to a registry database, the `operatorbundle` table schema must be updated to enable more than one mediatype. The table will be extended with a new `mediatype` column with a value that can be one of `{registry+v1, registry+v2}`. A respective migration will also be added to [operator-registy](https://github.com/operator-framework/operator-registry/tree/master/pkg/sqlite/migrations) to cover this schema change.
+
+#### Registry API Changes
+
+A new endpoint will be added to service requests for all clients that understand how to unpack `registry+v2` bundles. The previous endpoint will remain active, but will return errors for queries that result in `registry+v1` bundles. This will allow old clients to continue operating while preventing them from resolving bundles they cannot unpack. The errors -- which are typically bubbled to top-level resources -- will help to make unsupported bundle type errors more obvious to end users expecting an update to content managed by an older version of OLM.
```

Comment:
Should this say that the existing endpoint will return errors for queries that result in `registry+v2` bundles? Also, is that still true if a response includes bundles of multiple types (in that case, will it filter the results instead of failing)?

How does this approach compare to adding a content type negotiation mechanism to registry? It could be as basic as adding a field to requests that specifies a set of acceptable content types, for those RPCs that may return bundles.

Reply:
> Should this say that the existing endpoint will return errors for queries that result in registry+v2 bundles?

yup -- typo

> Also, is that still true if a response includes bundles of multiple types (in that case, will it filter the results instead of failing)?

It should probably return an error instead of filtering so the client knows something is up; i.e. silent failures are misleading.

> How does this approach compare to adding a content type negotiation mechanism to registry? It could be as basic as adding a field to requests that specifies a set of acceptable content types, for those RPCs that may return bundles.

It's probably less flexible than using an `accepts` field, but I'm not sure how that would work for existing clusters? Do we assume the absence of an `accepts` field means the client accepts only `registry+v1` bundles?

Reply:
> It should probably return an error instead of filtering so the client knows something is up; i.e. silent failures are misleading.

It might not be tenable to be so strict. For example, the ListBundles API will become effectively broken for old clients the moment a v2 bundle is added to an index.

> Do we assume the absence of an accepts field means the client accepts only registry+v1 bundles?

That's what I was imagining. WDYT?

Reply:
> the ListBundles API will become effectively broken for old clients the moment a v2 bundle is added to an index.

I thought that list-bundles only retrieved the key. 

Filtering probably won't be misleading if it applies to all queries (so even package server doesn't see new updates).

> That's what I was imagining. WDYT?

we can go this route, I just need to make sure adding fields to an RPC doesn't break API backwards compatibility.

Reply:
Updated!